### ISSUE TITLE
Add standalone Windows version (leproxy.exe)

### ISSUE THREAD

**Issue by clue** [leproxy/leproxy#59]
LeProxy runs on any platform that is supported by PHP, including Windows. On Windows, that means that users have to first download PHP and then run the ˋleproxy.phpˋ file on the command line. Arguably, this is not how many Windows users are used to run common applications.

As an alternative, we may want to bundle the ˋleproxy.phpˋ file with a PHP runtime in a single executable. This way, Windows users only have to download a single ˋleproxy.exeˋ that can be executed right away like a portable app.

**Comment by clue** [comment]
Here are some quick results from trying to build the current `leproxy-0.2.2.php` into a `leproxy-0.2.2.exe`:

ExeOutput: 8.432.789
- https://www.exeoutput.com/
- proprietary with 14 day trial, reasonable price tag
- works out-of-the-box, no visual studio runtime requirement
- annoying GUI / installation (which makes sense for all the features it has, but we're barely using 10% of this)
- hard to embed into automated build script

RapidEXE: 5.902.149
- https://deneskellner.com/sw/rapidexe by @DenesKellner
- exec missing vcruntime140.dll ([PHP homepage says](https://windows.php.net/): "The VC14, VC15 and VS16 builds require to have the Visual C++ Redistributable for Visual Studio 2015-2019 x64 or x86 installed")
- supports arbitrary runtimes, useful for future PHP upgrades

PHPFlexer: version n/a
- https://deneskellner.com/sw/phpflexer by @DenesKellner
- run missing vcruntime120.dll
- legacy version, superseded by RapidEXE

7zip: 3.139.449
- exec automatic?!
- most solutions seem to use 7zip SFX archives, e.g. https://github.com/islamadel/bat2exe/blob/master/source/archiv/bat2exe_drs.bat#L185

The gist is that there are a number of projects that allow building a portable executable for LeProxy that work to varying degrees. The goal would be to come up with a solution that requires little maintenance and something that we can automate so that the next time we tag a new LeProxy release, we should be able to easily generate a resulting executable for Windows. Stretch goal: In a perfect world, we would be able to invoke this build script without booting into a Windows environment first. If anybody feels like taking a look into this, any input would be much appreciated :+1: :tada:

**Comment by beppe9000** [comment]
have you explored other alternatives like php compilation to a dotnet assembly? e.g. the phalanger tool is open source

**Comment by clue** [comment]
> have you explored other alternatives like php compilation to a dotnet assembly? e.g. the phalanger tool is open source

No I haven't, but I'm open for input in case anybody feels like taking a look at this :+1: 

I'm currently evaluating commercial options and seeing if it makes sense to provide some kind of paid / pro version which could potentially include the portable Windows application version.

Again, any input is welcome :+1:

**Comment by beppe9000** [comment]
I think a winrar/similar self extracting archive that dumps leproxy and a php package in temp and runs it passing arguments is probably more appropriate.

<hr>

While I think real exe compilation might be an overkill (to properly transpile/compile with static linking), I investigated the [PeachiePie compiler](https://github.com/peachpiecompiler), an evolution of phalanger and [tested it](https://gist.github.com/beppe9000/f73b16c3b13841047583b6340f2e4ebb) successfully with some simple scripts. It seem not to cover the entire php api surface, but they have a [detailed tabulation](https://docs.peachpie.io/compatibility-status/) of what's there and what is not. You surely have more knowledge of the code base and could make an asessment of that. Maybe a list of things that peahpie must support to make leproxy work with it.

**Comment by clue** [comment]
I'm not a Windows user myself, so this isn't a high-priority issue for me. That's why I'm currently leaning towards evaluating a paid / pro version first to see if there's any interest before spending too much time on this :+1:

**Comment by DenesKellner** [comment]
Hi, I'm the author of RapidEXE.
Is that runtime dll the only thing you're missing?

**Comment by beppe9000** [comment]
> Hi, I'm the author of RapidEXE.

Is RapidEXE open source? I found a reference to a compile error on the website but I can't find the sources...

@clue the peachpie project can leverage the cross-compile function of .net core

**Comment by DenesKellner** [comment]
> Is RapidEXE open source? 
No, not _yet_. But that's just because I had no reason to make it so; it can easily become open source tomorrow.

> I found a reference to a compile error on the website but I can't find the sources...
RapidEXE itself is working properly; the exe it creates may have errors. Which problem are you referring to? The website describes a few cases in the troubleshooting section, do you mean one of those?

**Comment by beppe9000** [comment]
I probably fell victim of a context ambiguity while checking out the website: I didn't know if "compile" referred to compiling rapidexe itself or the process of bundling the packaged app.  Now it's clear it applies to the latter. 

> No, not _yet_. But that's just because I had no reason to make it so; it can easily become open source tomorrow.

Great! I look forward to have fully transparent toolchain 👍 
Also I was thinking about creating something similar for working with portable applications, but it's better to improve the wheel rather than to re-invent it....

**Comment by DenesKellner** [comment]
> I probably fell victim of a context ambiguity while checking out the website: I didn't know if "compile" referred to compiling rapidexe itself or the process of bundling the packaged app. Now it's clear it applies to the latter.

Exactly.

> I was thinking about creating something similar for working with portable applications, but it's better to improve the wheel rather than to re-invent it....

That's one good reason to make it open source!
In fact I'm working on RX3 right as we speak so maybe the Way of the Samurai would be to move this project to the global space and let everyone add their ideas to the wheel.

Hmm. Worth considering it is. Decide it wisely I must.
